File: 40. Combination Sum II/40. Combination Sum II.go
func combinationSum2(candidates []int, target int) [][]int {
    sort.Ints(candidates)
    var result [][]int
    backtrack(candidates, target, 0, []int{}, &result)
    return result
}

func backtrack(candidates []int, target int, index int, current []int, result *[][]int) {
    if target == 0 {
        temp := make([]int, len(current))
        copy(temp, current)
        *result = append(*result, temp)
        return
    }
    
    for i := index; i < len(candidates); i++ {
        if i > index && candidates[i] == candidates[i-1] {
            continue
        }
        if candidates[i] > target {
            break
        }
        current = append(current, candidates[i])
        backtrack(candidates, target - candidates[i], i + 1, current, result)
        current = current[:len(current)-1]
    }
}

File: 40. Combination Sum II/40. Combination Sum II.cs
public class Solution {
    public IList<IList<int>> CombinationSum2(int[] candidates, int target) {
        Array.Sort(candidates);
        var result = new List<IList<int>>();
        Backtrack(candidates, target, 0, new List<int>(), result);
        return result;
    }
    
    private void Backtrack(int[] candidates, int target, int index, List<int> current, List<IList<int>> result) {
        if (target == 0) {
            result.Add(new List<int>(current));
            return;
        }
        
        for (int i = index; i < candidates.Length; i++) {
            if (i > index && candidates[i] == candidates[i-1]) continue;
            if (candidates[i] > target) break;
            current.Add(candidates[i]);
            Backtrack(candidates, target - candidates[i], i + 1, current, result);
            current.RemoveAt(current.Count - 1);
        }
    }
}